client/src/main/java/io/pravega/client/batch/impl/BatchClientImpl.java

```diff
@@ -60,43 +68,98 @@ private StreamInfo getStreamInfo(Stream stream) {
+
+        CompletableFuture<Set<Segment>> segmentsFuture =
+                CompletableFuture.allOf(startStreamCut, endStreamCut)
+                                 .thenCompose(v -> {
+                                     StreamCut startSC = startStreamCut.join();
+                                     StreamCut endSC = endStreamCut.join();
+                                     log.debug("Start stream cut: {}, End stream cut: {}", startSC, endSC);
+                                     StreamSegmentInfo.validateStreamCuts(startSC, endSC);
+                                     return controller.getSegments(startSC, endSC);
+                                 });
+
+        final SortedSet<Segment> segmentSet = new TreeSet<>();
+        segmentSet.addAll(getAndHandleExceptions(segmentsFuture, RuntimeException::new));
+        log.debug("List of Segments between the start and end stream cuts : {}", segmentSet);
+        Iterator<SegmentInputSplit> iterator = Iterators.transform(segmentSet.iterator(),
+                s -> getSegmentInputSplit(s, startStreamCut.join(), endStreamCut.join()));
+        return StreamSegmentInfo.builder().segmentInputSplitIterator(iterator)
+                                .startStreamCut(startStreamCut.join())
+                                .endStreamCut(endStreamCut.join()).build();
+    }
+
+    private SegmentInfo getSegmentInfo(final Segment s) {
         @Cleanup
-        SegmentMetadataClient metadataClient = segmentMetadataClientFactory.createSegmentMetadataClient(segment);
-        SegmentInfo segmentInfo = metadataClient.getSegmentInfo();
-        return new SegmentIteratorImpl<>(inputStreamFactory, segment, deserializer, startingOffset, segmentInfo.getWriteOffset());
+        SegmentMetadataClient client = segmentMetadataClientFactory.createSegmentMetadataClient(s);
+        return client.getSegmentInfo();
     }
 
+    /*
+     * Given a segment fetch its SegmentInputSplit.
+     * - If segment is part of startStreamCut / endStreamCut update startOffset and endOffset accordingly.
+     * - If segment is not part of the streamCuts fetch the data using SegmentMetadataClient.
+     */
+    private SegmentInputSplit getSegmentInputSplit(final Segment segment, final StreamCut startStreamCut,
+                                                   final StreamCut endStreamCut) {
+        SegmentInputSplit.SegmentInputSplitBuilder segmentSplitBuilder = SegmentInputSplit.builder().segment(segment);
+        if (startStreamCut.getPositions().containsKey(segment) && endStreamCut.getPositions().containsKey(segment)) {
+            //use the meta data present in startStreamCut and endStreamCuts.
+            segmentSplitBuilder.startOffset(startStreamCut.getPositions().get(segment))
+                           .endOffset(endStreamCut.getPositions().get(segment));
+        } else {
+            //use segment meta data client to fetch the segment offsets.
+            SegmentInfo r = getSegmentInfo(segment);
```

Comment:
Maybe the per-segment remote call could be avoided by using sentinel values.  Do you really need to know the offsets of the intermediate segments?  The user-supplied cuts are still needed, of course.

Reply:
yes, the length is used by the SegmentIteratorImpl#hasNext()